File: 2019/python/day16.py
def part1(signal):
    lst = signal[:]
    length = len(lst)
    for _ in range(100):
        prev = lst[:]
        for i in range(length//2+1):
            d, total = i + 1, 0
            n = i
            while n < length:
                total += sum(prev[n : n+d])
                n += d*4
            n = i + d*2
            while n < length:
                total -= sum(prev[n : n+d])
                n += d*4
            lst[i] = abs(total) % 10
        cumsum = 0
        for i in range(length - 1, length//2, -1):
            cumsum += lst[i]
            lst[i] = cumsum % 10

    return ''.join(str(i) for i in lst[:8])

def part2(signal, offset):
    lst = (signal * 10_000)[offset:]
    length = len(lst)
    for _ in range(100):
        cumsum = 0
        for i in range(length - 1, -1, -1):
            cumsum += lst[i]
            lst[i] = cumsum % 10
    return ''.join(str(i) for i in lst[:8])


with open('../day_16.txt', 'r') as f:
    input_signal = f.read().strip()
    signal = [int(c) for c in input_signal]

print('part 1:', part1(signal))
print('part 2:', part2(signal, int(input_signal[0:7])))

File: 2019/go/day16.go
package main

import (
	"AdventOfCode/util"
	"fmt"
	"io/ioutil"
	"strconv"
)

func getSignal(data []byte) []int {
	signal := make([]int, len(data))
	for i, b := range data {
		signal[i] = int(b - '0')
	}
	return signal
}

func part1(signal []int, phases int) []int {
	l := len(signal)
	for p := 0; p < phases; p++ {
		next := make([]int, l)
		for i := 0; i < l; i++ {
			j, d, total := i, i+1, 0
			for j < l {
				for k := j; k < j+d && k < l; k++ {
					total += signal[k]
				}
				j += d + d
				for k := j; k < j+d && k < l; k++ {
					total -= signal[k]
				}
				j += d + d
			}
			next[i] = util.Abs(total) % 10
		}
		signal = next
	}
	return signal[:8]
}

func part2(signal []int, phases int, offset int) []int {
	l := len(signal)
	realSignal := make([]int, l*10000-offset)
	rl := len(realSignal)

	n := rl - l
	for ; n > 0; n -= l {
		copy(realSignal[n:n+l], signal)
	}
	copy(realSignal[0:n+l], signal[-n:])

	for p := 0; p < phases; p++ {
		sum := 0
		for i := rl - 1; i >= 0; i-- {
			sum += realSignal[i]
			realSignal[i] = sum % 10
		}
	}
	return realSignal[:8]
}

func main() {
	input, _ := ioutil.ReadFile("../day_16.txt")
	signal := getSignal(input[0 : len(input)-2])

	fmt.Printf("part 1: %v\n", part1(signal, 100))
	offset, _ := strconv.Atoi(string(input[0:7]))
	fmt.Printf("part 2: %v\n", part2(signal, 100, offset))
}
